presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java

```diff
@@ -804,6 +815,116 @@ public void testSchemaEvolution()
+
+            assertMultiTableFilterProject("discount > 0.01", "COUNT(*)");
+            assertMultiTableFilterProject("discount > 0.01 and discount + tax > 0.03", "COUNT(*)");
+            assertMultiTableFilterProject("discount = 0.0", "COUNT(*)");
+
+            assertMultiTableFilterProject("discount_real > 0.01", "COUNT(*)");
+            assertMultiTableFilterProject("tax_real > 0.01 and discount_real > 0.01", "COUNT(*)");
+
+            assertMultiTableFilterProject("keys IS NOT NULL", "keys");
+            assertMultiTableFilterProject("keys IS NULL", "keys");
+            assertMultiTableFilterProject("keys[1] % 5 = 0 AND keys[2] > 100", "orderkey");
+
+            assertMultiTableFilterProject("is_returned=false", "COUNT(*)");
+            assertMultiTableFilterProject("is_returned is NULL", "COUNT(*)");
+
+            assertMultiTableFilterProject("ship_day > 2", "ship_day");
+
+            assertMultiTableFilterProject("discount_long_decimal > 0.05", "discount_long_decimal");
+            assertMultiTableFilterProject("tax_short_decimal < 0.03", "tax_short_decimal");
+            assertMultiTableFilterProject("discount_long_decimal > 0.01 AND tax_short_decimal > 0.01 AND (discount_long_decimal + tax_short_decimal) < 0.03", "discount_long_decimal");
+
+            assertMultiTableFilterProject("dates[1].day % 2 = 0", "dates");
+            assertMultiTableFilterProject("info IS NOT NULL AND dates IS NOT NULL AND info.orderkey % 7 = 0", "info.orderkey, dates ");
+
+            // empty result
+            assertQueryReturnsEmptyResult("SELECT comment FROM test_file_format_orc WHERE orderkey < 0");
+            assertQueryReturnsEmptyResult("SELECT comment FROM test_file_format_orc WHERE comment LIKE '???'");
+        }
+        finally {
+            assertUpdate("DROP TABLE IF EXISTS test_file_format");
+            assertUpdate("DROP TABLE test_file_format_orc");
+        }
+    }
+
+    private Path getPartitionDirectory(String tableName, String partitionClause)
+    {
+        String filePath = ((String) computeActual(noPushdownFilter(getSession()), format("SELECT \"$path\" FROM %s WHERE %s LIMIT 1", tableName, partitionClause)).getOnlyValue())
+                .replace("file:", "");
+        return Paths.get(filePath).getParent();
+    }
+
+    private void assertMultiTableFilterProject(String filter, String project, String table1, String table2)
+    {
+        if (filter == null || filter.trim().isEmpty()) {
+            filter = "1=1";
+        }
+        assertEqualsIgnoreOrder(computeActual(format("SELECT * FROM %s WHERE %s", table1, filter)), computeActual(format("SELECT * FROM %s WHERE %s", table2, filter)));
+        assertEqualsIgnoreOrder(
+                computeActual(format("SELECT %s FROM %s WHERE %s", project, table1, filter)),
+                computeActual(format("SELECT %s FROM %s WHERE %s", project, table2, filter)));
+    }
+
+    private void assertMultiTableFilterProject(String filter, String project)
```

Comment:
method name is quite generic, but implementation is very specific; let's use `assertQueryUsingH2Cte`